File: src/workflow/results.ts
/**
 * Step-result contract: the typed shape of every natural step result, the field names
 * derived from those shapes, and the verdict parse shared by the runner and `hwf response check`.
 */

/** Blocking local command result. */
export type CommandResult = {
  stdout: string;
  stderr: string;
  exit_code: number;
  failed: boolean;
};

/** Blocking managed agent turn result. `verdict` is present only when the step declares `expect`. */
export type AgentResult = {
  response: string;
  agent: Record<string, unknown>;
  pane_id: string;
  verdict?: string;
};

/** Identifiers a placed command adds to the native readiness payload. */
export type ReadinessIds = {
  pane_id: string;
  tab_id: string;
  workspace_id: string;
};

/** Diagnostic keys a failed step carries into history through the failure fact. */
export type StepFailureDetails = {
  stdout?: string;
  stderr?: string;
  exit_code?: number;
  method?: string;
  verdict?: string;
};

type FieldNames<T> = Record<keyof T, true>;

const COMMAND_KEYS: FieldNames<CommandResult> = {
  stdout: true,
  stderr: true,
  exit_code: true,
  failed: true,
};

const AGENT_KEYS: FieldNames<Required<AgentResult>> = {
  response: true,
  agent: true,
  pane_id: true,
  verdict: true,
};

const READINESS_KEYS: FieldNames<ReadinessIds> = {
  pane_id: true,
  tab_id: true,
  workspace_id: true,
};

/** Native herdr AgentInfo, addressed through the herdr result-path table rather than field names. */
export const AGENT_INFO_FIELD = "agent" satisfies keyof AgentResult;

/** Only readable when the producing step declares `expect`. */
export const AGENT_VERDICT_FIELD = "verdict" satisfies keyof AgentResult;

export const COMMAND_EXIT_CODE_FIELD = "exit_code" satisfies keyof CommandResult;

export type ScalarFieldType = "string" | "number" | "boolean";

/** Scalar type of each command-result field, single-sourced beside the field-name sets. */
const COMMAND_FIELD_TYPE_MAP = {
  stdout: "string",
  stderr: "string",
  [COMMAND_EXIT_CODE_FIELD]: "number",
  failed: "boolean",
} satisfies Record<keyof CommandResult, ScalarFieldType>;

export const COMMAND_FIELD_TYPES: ReadonlyMap<string, ScalarFieldType> = new Map(
  Object.entries(COMMAND_FIELD_TYPE_MAP),
);

export const COMMAND_FIELDS: ReadonlySet<string> = new Set(Object.keys(COMMAND_KEYS));

export const AGENT_STRING_FIELDS: ReadonlySet<string> = new Set(
  Object.keys(AGENT_KEYS).filter(
    (name) => name !== AGENT_INFO_FIELD && name !== AGENT_VERDICT_FIELD,
  ),
);

export const READINESS_ID_FIELDS: ReadonlySet<string> = new Set(Object.keys(READINESS_KEYS));

/** Context keys a transcript can reach, which `returns:` must not carry out of the run. */
export const SENSITIVE_CONTEXT_KEYS: ReadonlySet<string> = new Set([
  "transcript",
  "transcript_file",
]);

function shapeOf(fields: Iterable<string>): string {
  return `{${[...fields].join(", ")}}`;
}

const AGENT_FIELD_ORDER = [
  "response",
  AGENT_INFO_FIELD,
  "pane_id",
] as const satisfies readonly (keyof AgentResult)[];

export const COMMAND_RESULT_SHAPE = shapeOf(COMMAND_FIELDS);

export const AGENT_RESULT_SHAPE = shapeOf(AGENT_FIELD_ORDER);

export const VERDICT_TOKEN_PATTERN = "[A-Z][A-Z0-9_]{0,31}";
export const VERDICT_TOKEN_RE = new RegExp(`^${VERDICT_TOKEN_PATTERN}$`);

export type ExpectSpec = {
  oneOf: string[];
  require?: string[];
};

export type VerdictParse = { ok: true; verdict: string } | { ok: false; line: string };

function finalNonEmptyLine(text: string): string {
  const lines = text.split("\n");
  for (let i = lines.length - 1; i >= 0; i--) {
    const line = lines[i]!.trim();
    if (line) return line;
  }
  return "";
}

/** The one verdict oracle: final non-empty line, trimmed, matched exactly against the tokens. */
export function parseVerdict(response: string, oneOf: readonly string[]): VerdictParse {
  const line = finalNonEmptyLine(response);
  if (oneOf.includes(line)) return { ok: true, verdict: line };
  return { ok: false, line };
}

export function verdictMismatchMessage(line: string, oneOf: readonly string[]): string {
  const found = line ? JSON.stringify(line) : "an empty response";
  return `final non-empty line ${found} is not a verdict token — expected exactly one of: ${oneOf.join(", ")}`;
}

export function verdictNotRequiredMessage(verdict: string, require: readonly string[]): string {
  return `verdict ${verdict} is not accepted — this step requires one of: ${require.join(", ")}`;
}

export type VerdictTokens = { ok: true; tokens: string[] } | { ok: false; error: string };

/** Decode a comma-separated `--one-of` list under the same rules as `expect.one_of`. */
export function parseVerdictTokens(raw: string): VerdictTokens {
  const tokens = raw
    .split(",")
    .map((token) => token.trim())
    .filter((token) => token.length > 0);
  if (tokens.length === 0) {
    return { ok: false, error: "--one-of requires at least one verdict token" };
  }
  for (const token of tokens) {
    if (!VERDICT_TOKEN_RE.test(token)) {
      return {
        ok: false,
        error: `invalid verdict token '${token}' — must match ${VERDICT_TOKEN_PATTERN}`,
      };
    }
  }
  const seen = new Set<string>();
  for (const token of tokens) {
    if (seen.has(token)) return { ok: false, error: `duplicate verdict token '${token}'` };
    seen.add(token);
  }
  return { ok: true, tokens };
}

File: internal/workflow/results.go
package workflow

import (
	"bytes"
	"encoding/json"
	"fmt"
	"strings"
)

// jsonQuote renders a string exactly as JavaScript JSON.stringify does.
// The observable echo surface uses this escaping, not Go strconv.Quote.
func jsonQuote(s string) string {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	_ = enc.Encode(s)
	return strings.TrimSuffix(buf.String(), "\n")
}

// CommandResult is the natural result of a blocking local command.
type CommandResult struct {
	Stdout   string `json:"stdout"`
	Stderr   string `json:"stderr"`
	ExitCode int    `json:"exit_code"`
	Failed   bool   `json:"failed"`
}

// AgentResult is the natural result of a blocking managed agent turn.
type AgentResult struct {
	Response string         `json:"response"`
	Agent    map[string]any `json:"agent"`
	PaneID   string         `json:"pane_id"`
	Verdict  string         `json:"verdict,omitempty"`
}

// ReadinessResult is the native wait result plus identifiers that the
// placed command creates.
type ReadinessResult struct {
	PaneID      string `json:"pane_id"`
	TabID       string `json:"tab_id"`
	WorkspaceID string `json:"workspace_id"`
}

// StepFailureDetails carries command and verdict diagnostics.
type StepFailureDetails struct {
	Stdout   string `json:"stdout,omitempty"`
	Stderr   string `json:"stderr,omitempty"`
	ExitCode *int   `json:"exit_code,omitempty"`
	Method   string `json:"method,omitempty"`
	Verdict  string `json:"verdict,omitempty"`
}

const (
	AgentInfoField    = "agent"
	AgentVerdictField = "verdict"
	CommandExitField  = "exit_code"
)

var (
	CommandFields = map[string]bool{
		"stdout": true, "stderr": true, "exit_code": true, "failed": true,
	}
	AgentStringFields = map[string]bool{
		"response": true, "pane_id": true,
	}
	ReadinessIDFields = map[string]bool{
		"pane_id": true, "tab_id": true, "workspace_id": true,
	}
	SensitiveContextKeys = map[string]bool{
		"transcript": true, "transcript_file": true,
	}
	CommandFieldTypes = map[string]string{
		"stdout": "string", "stderr": "string", "exit_code": "number", "failed": "boolean",
	}
)

// ParseVerdict applies the one verdict rule that the runner and CLI share.
func ParseVerdict(response string, oneOf []string) (string, bool, string) {
	lines := strings.Split(response, "\n")
	line := ""
	for i := len(lines) - 1; i >= 0; i-- {
		if trimmed := strings.TrimSpace(lines[i]); trimmed != "" {
			line = trimmed
			break
		}
	}
	for _, token := range oneOf {
		if token == line {
			return line, true, ""
		}
	}
	return "", false, line
}

func VerdictMismatchMessage(line string, oneOf []string) string {
	found := "an empty response"
	if line != "" {
		found = jsonQuote(line)
	}
	return fmt.Sprintf("final non-empty line %s is not a verdict token — expected exactly one of: %s", found, strings.Join(oneOf, ", "))
}

func VerdictNotRequiredMessage(verdict string, required []string) string {
	return fmt.Sprintf("verdict %s is not accepted — this step requires one of: %s", verdict, strings.Join(required, ", "))
}

// ParseVerdictTokens decodes the CLI comma-separated verdict option.
func ParseVerdictTokens(raw string) ([]string, error) {
	var tokens []string
	for _, token := range strings.Split(raw, ",") {
		token = strings.TrimSpace(token)
		if token == "" {
			continue
		}
		tokens = append(tokens, token)
	}
	if len(tokens) == 0 {
		return nil, fmt.Errorf("--one-of requires at least one verdict token")
	}
	for _, token := range tokens {
		if !verdictTokenRE.MatchString(token) {
			return nil, fmt.Errorf("invalid verdict token '%s' — must match %s", token, VerdictTokenPattern)
		}
	}
	seen := make(map[string]bool)
	for _, token := range tokens {
		if seen[token] {
			return nil, fmt.Errorf("duplicate verdict token '%s'", token)
		}
		seen[token] = true
	}
	return tokens, nil
}
